browser/search_engines/android/java/src/org/chromium/brave/browser/search_engines/settings/CustomSearchEngineAdapter.java

```diff
@@ -0,0 +1,72 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+package org.chromium.brave.browser.search_engines.settings;
+
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.DiffUtil;
+import androidx.recyclerview.widget.ListAdapter;
+
+import org.chromium.chrome.browser.search_engines.R;
+
+public class CustomSearchEngineAdapter extends ListAdapter<String, CustomSearchEngineViewHolder> {
+    private CustomSearchEnginesCallback mCustomSearchEnginesCallback;
+
+    protected CustomSearchEngineAdapter() {
+        super(DIFF_CALLBACK);
+    }
+
+    private static final DiffUtil.ItemCallback<String> DIFF_CALLBACK =
+            new DiffUtil.ItemCallback<String>() {
+                @Override
+                public boolean areItemsTheSame(@NonNull String oldItem, @NonNull String newItem) {
+                    return oldItem.equals(newItem);
+                }
+
+                @Override
+                public boolean areContentsTheSame(
+                        @NonNull String oldItem, @NonNull String newItem) {
+                    return oldItem.equals(newItem);
+                }
+            };
+
+    public void setCustomSearchEnginesCallback(
+            CustomSearchEnginesCallback customSearchEnginesCallback) {
+        mCustomSearchEnginesCallback = customSearchEnginesCallback;
+    }
+
+    @Override
+    public void onBindViewHolder(
+            @NonNull CustomSearchEngineViewHolder customSearchEngineViewHolder, int position) {
+        final String searchEngineKeyword = getItem(position);
+
+        customSearchEngineViewHolder.getSearchEngineText().setText(searchEngineKeyword);
+
+        customSearchEngineViewHolder
```

Comment:
I don't understand what is going on here at all, this adapter depends on the fragment? Why would it do that? That's completely backwards from how `SearchEngineAdapter` works in upstream with `SearchEngineSettings`

Reply:
We need to open a fragment to allow user to edit the search engine on click event. here : https://github.com/brave/brave-core/blob/4219afbf311b4a162c4d7d0ff8e004f73632f553/browser/search_engines/android/java/src/org/chromium/brave/browser/search_engines/settings/CustomSearchEnginesPreference.java#L78

Reply:
The fragment already owns the adapter, why do we need to pass the fragment back to the adapter again as this Callback interface?

Reply:
we are not passing the fragment or it's context. we are passing the implementation of the interface which performs actions for individual item.

Reply:
sorry, it's a Preference, not a Fragment, but the issue is the same because it owns the adapter

Reply:
the general point here is that this is a very unusual pattern and seems unnecessarily complicated. I don't see this kind of thing in other UIs in upstream.